src/ncp/ncp_base.hpp

```diff
@@ -780,6 +805,54 @@ class NcpBase
 #endif
 
 #if OPENTHREAD_CONFIG_NCP_DNSSD_ENABLE && OPENTHREAD_CONFIG_PLATFORM_DNSSD_ENABLE
+
+    struct DnssdRequest
+    {
+        otPlatDnssdRegisterCallback mCallback;
+        otPlatDnssdRequestId        mId;
+        bool                        mIsInUse;
+    };
```

Comment:
Just realized this: The current model may not work correctly.

Using the request ID to track an entry in platform is actually not correct. At platform implementation,  for `otPlatDnssd` API,  we cannot assume `aRequestId` will be different/unique, as mentioned in the API documentation:
https://github.com/openthread/openthread/blob/b0984aff7a64ab8c0d745fdde741ded77e95ffa7/include/openthread/platform/dnssd.h#L198-L201

The "Request ID" is primarily for the OT stack's use itself, allowing it to distinguish different registration.
Within the platform layer, we should remember both the "request ID and callback" together (sort of like a callback and its `aContext` but instead of `void *aContext` we have a numerical `aRequestId` instead).

Reply:
Thanks for finding the problem! Then I think we should maintain both request ID and the callback to track a request. And we also add the pointer to the callback as a data blob field nto the spinel property.

I think of another problem: In some cases there could be two requests that have exactly the same context. Same requests and same callback (because it is just a static function in OT). For example, if AdvProxy does 2 `RegisterService` with the same request ID, we will be in that situation. In this case, when we do match, we can match more than one entry. Does it matter that we choose which matched entry since the request Id and callbacks are exactly the same? (That means we can simply find the first matched entry can call the callback and clean). Or we won't go into such cases? @abtink

Reply:
> Then I think we should maintain both request ID and the callback to track a request. And we also add the pointer to the callback as a data blob field nto the spinel property.

I think we don't need a list in `ncp_base.cpp` to track requests. We can pass everything through the spinel property to the host (including the request ID and callback pointer). When we get the result back, we retrieve the callback and request ID from the spinel response and then invoke the callback.

Host names or service instance names can be used to track registrations. The callback and request ID simply provide a way for the platform to report the result of a request.

Reply:
Sounds good to me. Thanks!

Reply:
Updated, thanks! The code becomes simpler without maintaining the list.